File: mini-claw-code-py/src/mini_claw_code_py/streaming.py
from __future__ import annotations

import asyncio
import json
from collections import deque
from dataclasses import dataclass
from typing import Any, Protocol, Sequence

from .agent import AgentDone, AgentError, AgentEvent, AgentTextDelta, AgentToolCall, tool_summary
from .mock import MockProvider
from .types import AssistantTurn, Message, StopReason, ToolCall, ToolDefinition, ToolSet


@dataclass(slots=True)
class TextDelta:
    text: str


@dataclass(slots=True)
class ToolCallStart:
    index: int
    id: str
    name: str


@dataclass(slots=True)
class ToolCallDelta:
    index: int
    arguments: str


@dataclass(slots=True)
class StreamDone:
    pass


StreamEvent = TextDelta | ToolCallStart | ToolCallDelta | StreamDone


@dataclass(slots=True)
class PartialToolCall:
    id: str = ""
    name: str = ""
    arguments: str = ""


class StreamAccumulator:
    def __init__(self) -> None:
        self.text = ""
        self.tool_calls: list[PartialToolCall] = []

    def feed(self, event: StreamEvent) -> None:
        if isinstance(event, TextDelta):
            self.text += event.text
        elif isinstance(event, ToolCallStart):
            while len(self.tool_calls) <= event.index:
                self.tool_calls.append(PartialToolCall())
            self.tool_calls[event.index].id = event.id
            self.tool_calls[event.index].name = event.name
        elif isinstance(event, ToolCallDelta):
            if event.index < len(self.tool_calls):
                self.tool_calls[event.index].arguments += event.arguments

    def finish(self) -> AssistantTurn:
        text = self.text or None
        tool_calls = [
            ToolCall(
                id=call.id,
                name=call.name,
                arguments=_decode_json(call.arguments),
            )
            for call in self.tool_calls
            if call.name
        ]
        stop_reason = StopReason.TOOL_USE if tool_calls else StopReason.STOP
        return AssistantTurn(text=text, tool_calls=tool_calls, stop_reason=stop_reason)


def parse_sse_line(line: str) -> list[StreamEvent] | None:
    if not line.startswith("data: "):
        return None

    data = line[len("data: ") :]
    if data == "[DONE]":
        return [StreamDone()]

    try:
        chunk = json.loads(data)
    except json.JSONDecodeError:
        return None

    choices = chunk.get("choices") or []
    if not choices:
        return None

    delta = (choices[0] or {}).get("delta") or {}
    events: list[StreamEvent] = []

    content = delta.get("content")
    if isinstance(content, str) and content:
        events.append(TextDelta(content))

    for tool_call in delta.get("tool_calls") or []:
        index = tool_call.get("index")
        if not isinstance(index, int):
            continue
        call_id = tool_call.get("id")
        function = tool_call.get("function") or {}
        if isinstance(call_id, str):
            events.append(
                ToolCallStart(
                    index=index,
                    id=call_id,
                    name=function.get("name") or "",
                )
            )
        arguments = function.get("arguments")
        if isinstance(arguments, str) and arguments:
            events.append(ToolCallDelta(index=index, arguments=arguments))

    return events or None


class StreamProvider(Protocol):
    async def stream_chat(
        self,
        messages: Sequence[Message],
        tools: Sequence[ToolDefinition],
        queue: "asyncio.Queue[StreamEvent]",
    ) -> AssistantTurn:
        ...


class MockStreamProvider(StreamProvider):
    def __init__(self, responses: deque[AssistantTurn]) -> None:
        self.inner = MockProvider(responses)

    async def stream_chat(
        self,
        messages: Sequence[Message],
        tools: Sequence[ToolDefinition],
        queue: "asyncio.Queue[StreamEvent]",
    ) -> AssistantTurn:
        turn = await self.inner.chat(messages, tools)
        if turn.text:
            for char in turn.text:
                await queue.put(TextDelta(char))
        for index, call in enumerate(turn.tool_calls):
            await queue.put(ToolCallStart(index=index, id=call.id, name=call.name))
            await queue.put(ToolCallDelta(index=index, arguments=json.dumps(call.arguments)))
        await queue.put(StreamDone())
        return turn


class StreamingAgent:
    def __init__(self, provider: StreamProvider) -> None:
        self.provider = provider
        self.tools = ToolSet()

    def tool(self, tool: object) -> "StreamingAgent":
        self.tools.push(tool)  # type: ignore[arg-type]
        return self

    async def run(
        self,
        prompt: str,
        events: "asyncio.Queue[AgentEvent]",
    ) -> str:
        return await self.chat([Message.user(prompt)], events)

    async def chat(
        self,
        messages: list[Message],
        events: "asyncio.Queue[AgentEvent]",
    ) -> str:
        defs = self.tools.definitions()

        while True:
            stream_queue: asyncio.Queue[StreamEvent] = asyncio.Queue()

            async def forward() -> None:
                while True:
                    event = await stream_queue.get()
                    if isinstance(event, TextDelta):
                        await events.put(AgentTextDelta(event.text))
                    if isinstance(event, StreamDone):
                        return

            forwarder = asyncio.create_task(forward())

            try:
                turn = await self.provider.stream_chat(messages, defs, stream_queue)
            except Exception as exc:
                await events.put(AgentError(str(exc)))
                forwarder.cancel()
                raise

            await forwarder

            if turn.stop_reason is StopReason.STOP:
                text = turn.text or ""
                await events.put(AgentDone(text))
                messages.append(Message.assistant(turn))
                return text

            results: list[tuple[str, str]] = []
            for call in turn.tool_calls:
                await events.put(AgentToolCall(call.name, tool_summary(call)))
                tool = self.tools.get(call.name)
                if tool is None:
                    content = f"error: unknown tool `{call.name}`"
                else:
                    try:
                        content = await tool.call(call.arguments)
                    except Exception as exc:
                        content = f"error: {exc}"
                results.append((call.id, content))

            messages.append(Message.assistant(turn))
            for call_id, content in results:
                messages.append(Message.tool_result(call_id, content))


def _decode_json(raw: str) -> Any:
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return None

File: mini-claw-code-ts/src/streaming.ts
import { assistantMessage, toolResultMessage, ToolSet, userMessage, type AssistantTurn, type JsonValue, type Message, type ToolCall, type ToolDefinition } from "./types";
import type { AgentEventHandler } from "./agent";
import { toolSummary } from "./agent";
import type { Provider } from "./types";

export type StreamEvent =
  | { kind: "text_delta"; text: string }
  | { kind: "tool_call_start"; index: number; id: string; name: string }
  | { kind: "tool_call_delta"; index: number; arguments: string }
  | { kind: "done" };

export type StreamEventHandler = (event: StreamEvent) => void | Promise<void>;

export interface StreamProvider {
  streamChat(
    messages: Message[],
    tools: ToolDefinition[],
    onEvent: StreamEventHandler,
  ): Promise<AssistantTurn>;
}

interface PartialToolCall {
  id: string;
  name: string;
  arguments: string;
}

export class StreamAccumulator {
  #text = "";
  #toolCalls: PartialToolCall[] = [];

  feed(event: StreamEvent): void {
    switch (event.kind) {
      case "text_delta":
        this.#text += event.text;
        return;
      case "tool_call_start":
        while (this.#toolCalls.length <= event.index) {
          this.#toolCalls.push({ id: "", name: "", arguments: "" });
        }
        this.#toolCalls[event.index] = {
          ...this.#toolCalls[event.index]!,
          id: event.id,
          name: event.name,
        };
        return;
      case "tool_call_delta":
        this.#toolCalls[event.index] ??= { id: "", name: "", arguments: "" };
        this.#toolCalls[event.index]!.arguments += event.arguments;
        return;
      case "done":
        return;
    }
  }

  finish(): AssistantTurn {
    const toolCalls: ToolCall[] = this.#toolCalls
      .filter((toolCall) => toolCall.name.length > 0)
      .map((toolCall) => ({
        id: toolCall.id,
        name: toolCall.name,
        arguments: parseJsonOrNull(toolCall.arguments),
      }));

    return {
      text: this.#text.length > 0 ? this.#text : undefined,
      toolCalls,
      stopReason: toolCalls.length > 0 ? "tool_use" : "stop",
    };
  }
}

export function parseSseLine(line: string): StreamEvent[] | undefined {
  if (!line.startsWith("data: ")) {
    return undefined;
  }

  const data = line.slice(6);
  if (data === "[DONE]") {
    return [{ kind: "done" }];
  }

  const chunk = parseJsonOrNull(data) as {
    choices?: Array<{
      delta?: {
        content?: string;
        tool_calls?: Array<{
          index: number;
          id?: string;
          function?: { name?: string; arguments?: string };
        }>;
      };
    }>;
  } | null;

  const delta = chunk?.choices?.[0]?.delta;
  if (!delta) {
    return undefined;
  }

  const events: StreamEvent[] = [];

  if (delta.content) {
    events.push({ kind: "text_delta", text: delta.content });
  }

  for (const call of delta.tool_calls ?? []) {
    if (call.id) {
      events.push({
        kind: "tool_call_start",
        index: call.index,
        id: call.id,
        name: call.function?.name ?? "",
      });
    }
    if (call.function?.arguments) {
      events.push({
        kind: "tool_call_delta",
        index: call.index,
        arguments: call.function.arguments,
      });
    }
  }

  return events.length > 0 ? events : undefined;
}

export class MockStreamProvider implements StreamProvider {
  readonly #provider: Provider;

  constructor(provider: Provider) {
    this.#provider = provider;
  }

  async streamChat(
    messages: Message[],
    tools: ToolDefinition[],
    onEvent: StreamEventHandler,
  ): Promise<AssistantTurn> {
    const turn = await this.#provider.chat(messages, tools);

    for (const character of turn.text ?? "") {
      await onEvent({ kind: "text_delta", text: character });
    }

    for (const [index, call] of turn.toolCalls.entries()) {
      await onEvent({
        kind: "tool_call_start",
        index,
        id: call.id,
        name: call.name,
      });
      await onEvent({
        kind: "tool_call_delta",
        index,
        arguments: JSON.stringify(call.arguments),
      });
    }

    await onEvent({ kind: "done" });
    return turn;
  }
}

export class StreamingAgent<P extends StreamProvider> {
  readonly provider: P;
  readonly tools: ToolSet;

  constructor(provider: P, tools = new ToolSet()) {
    this.provider = provider;
    this.tools = tools;
  }

  static new<P extends StreamProvider>(provider: P): StreamingAgent<P> {
    return new StreamingAgent(provider);
  }

  tool(tool: Parameters<ToolSet["push"]>[0]): this {
    this.tools.push(tool);
    return this;
  }

  async run(prompt: string, onEvent?: AgentEventHandler): Promise<string> {
    const messages = [userMessage(prompt)];
    return this.chat(messages, onEvent);
  }

  async chat(messages: Message[], onEvent?: AgentEventHandler): Promise<string> {
    const definitions = this.tools.definitions();

    for (;;) {
      const turn = await this.provider.streamChat(messages, definitions, async (event) => {
        if (event.kind === "text_delta") {
          await onEvent?.({ kind: "text_delta", text: event.text });
        }
      });

      if (turn.stopReason === "stop") {
        const text = turn.text ?? "";
        messages.push(assistantMessage(turn));
        await onEvent?.({ kind: "done", text });
        return text;
      }

      const results: Array<[string, string]> = [];
      for (const call of turn.toolCalls) {
        await onEvent?.({
          kind: "tool_call",
          name: call.name,
          summary: toolSummary(call),
        });
        const tool = this.tools.get(call.name);
        const content = tool
          ? await tool.call(call.arguments).catch((error) => `error: ${toErrorMessage(error)}`)
          : `error: unknown tool \`${call.name}\``;
        results.push([call.id, content]);
      }

      messages.push(assistantMessage(turn));
      for (const [id, content] of results) {
        messages.push(toolResultMessage(id, content));
      }
    }
  }
}

function parseJsonOrNull(input: string): JsonValue {
  try {
    return JSON.parse(input) as JsonValue;
  } catch {
    return null;
  }
}

function toErrorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
